ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11blend.cpp

```diff
@@ -18,34 +18,72 @@
 #include "Bridges/General/DX11Context.h"
 #include "Direct3D11Headers.h"
 #include "Graphics_Systems/General/GSblend.h"
+#include "Graphics_Systems/General/GSprimitives.h"
 
 namespace enigma
 {
-  extern int currentblendmode[2];
-  extern int currentblendtype;
+
+extern int currentblendmode[2];
+extern int currentblendtype;
+
+} // namespace enigma
+
+using namespace enigma;
+
+namespace {
+
+D3D11_BLEND_DESC blendStateDesc = { };
+
+void update_blend_state() {
+  static ID3D11BlendState* pBlendState = NULL;
+  if (pBlendState) { pBlendState->Release(); pBlendState = NULL; }
+  m_device->CreateBlendState(&blendStateDesc, &pBlendState);
+  draw_batch_flush(batch_flush_deferred);
+  m_deviceContext->OMSetBlendState(pBlendState, NULL, 0xffffffff);
 }
 
+} // namespace anonymous
+
 namespace enigma_user
 {
 
-int draw_set_blend_mode(int mode){
-
+int draw_set_blend_mode(int mode) {
+  const static D3D11_BLEND dest_modes[] = {D3D11_BLEND_INV_SRC_ALPHA,D3D11_BLEND_ONE,D3D11_BLEND_INV_SRC_COLOR,D3D11_BLEND_INV_SRC_COLOR};
+  const static D3D11_BLEND_OP blend_ops[] = {D3D11_BLEND_OP_ADD,D3D11_BLEND_OP_ADD,D3D11_BLEND_OP_SUBTRACT,D3D11_BLEND_OP_MAX};
+  blendStateDesc.RenderTarget[0].BlendEnable = TRUE;
+  blendStateDesc.RenderTarget[0].SrcBlendAlpha = blendStateDesc.RenderTarget[0].SrcBlend = (mode == bm_subtract) ? D3D11_BLEND_ZERO : D3D11_BLEND_SRC_ALPHA;
+  blendStateDesc.RenderTarget[0].DestBlendAlpha = blendStateDesc.RenderTarget[0].DestBlend = dest_modes[mode % 4];
+  blendStateDesc.RenderTarget[0].BlendOpAlpha = blendStateDesc.RenderTarget[0].BlendOp = blend_ops[mode % 4];
+  blendStateDesc.RenderTarget[0].RenderTargetWriteMask = D3D11_COLOR_WRITE_ENABLE_ALL;
+  update_blend_state();
+  return 0;
 }
 
-int draw_set_blend_mode_ext(int src, int dest){
-
+int draw_set_blend_mode_ext(int src, int dest) {
+  const static D3D11_BLEND blend_equivs[] = {
+    D3D11_BLEND_ZERO, D3D11_BLEND_ONE, D3D11_BLEND_SRC_COLOR, D3D11_BLEND_INV_SRC_COLOR, D3D11_BLEND_SRC_ALPHA,
+    D3D11_BLEND_INV_SRC_ALPHA, D3D11_BLEND_DEST_ALPHA, D3D11_BLEND_INV_DEST_ALPHA, D3D11_BLEND_DEST_COLOR,
+    D3D11_BLEND_INV_DEST_COLOR, D3D11_BLEND_SRC_ALPHA_SAT
+  };
+  blendStateDesc.RenderTarget[0].BlendEnable = TRUE;
```

Comment:
Actually, I am caching it with `blendStateDesc` but the reason I am resetting that is because line 35's zero initialization originally sets it to false.
`
D3D11_BLEND_DESC blendStateDesc = { };
`

I suppose I should just make a single initialization method so I don't have to do that.

Reply:
@JoshDreamland here: 50dec1380c72295a7341c1707893976658a19c87

We needed to do that anyway so that we can add `draw_set_color_write_enable` for the color write mask without affecting the actual blend state/op.